addons/certificate/models/key.py

```diff
@@ -0,0 +1,263 @@
+        if not isinstance(private_key, rsa.RSAPrivateKey):
+            raise UserError(_("Unsupported asymmetric cryptography algorithm. Currently supported: RSA."))
+
+        return private_key.decrypt(
+            message,
+            padding.OAEP(
+                mgf=padding.MGF1(algorithm=STR_TO_HASH[hashing_algorithm]),
+                algorithm=STR_TO_HASH[hashing_algorithm],
+                label=None
+            )
+        )
+
+    @api.model
+    def _sign_with_key(self, message, pem_key, pwd=None, hashing_algorithm='sha256', block=True):
+        """ Return the base64 encoded signature of message. """
+
+        if not isinstance(message, bytes):
+            message = message.encode('utf-8')
+
+        if hashing_algorithm not in STR_TO_HASH:
+            raise UserError(_("Unsupported hashing algorithm. Currently supported: sha1 and sha256."))
+
+        try:
+            private_key = serialization.load_pem_private_key(base64.b64decode(pem_key), pwd)
+        except ValueError:
+            raise UserError(_("The private key could not be loaded."))
+
+        if isinstance(private_key, ec.EllipticCurvePrivateKey):
+            signature = private_key.sign(
+                message,
+                ec.ECDSA(STR_TO_HASH[hashing_algorithm])
+            )
+        elif isinstance(private_key, rsa.RSAPrivateKey):
+            signature = private_key.sign(
+                message,
+                padding.PKCS1v15(),
+                STR_TO_HASH[hashing_algorithm]
+            )
+        else:
+            raise UserError(_("Unsupported asymmetric cryptography algorithm. Currently supported: EC and RSA."))
+
+        if block:
+            return base64.encodebytes(signature)
+        else:
+            return base64.b64encode(signature)
+
+    @api.model
+    def _numbers_public_key_bytes_with_key(self, pem_key, block=True):
+        try:
+            public_key = serialization.load_pem_public_key(base64.b64decode(pem_key))
+        except ValueError:
+            raise UserError(_("The public key could not be loaded."))
+
+        if isinstance(public_key, ec.EllipticCurvePublicKey):
+            e = public_key.public_numbers().x
+            n = public_key.public_numbers().y
+        elif isinstance(public_key, rsa.RSAPublicKey):
+            e = public_key.public_numbers().e
+            n = public_key.public_numbers().n
+        else:
+            raise UserError(_("Unsupported asymmetric cryptography algorithm. Currently supported: EC, RSA."))
+
+        e = e.to_bytes((e.bit_length() + 7) // 8, 'big')
+        n = n.to_bytes((n.bit_length() + 7) // 8, 'big')
+
+        if block:
+            return base64.encodebytes(e), base64.encodebytes(n)
+        else:
+            return base64.b64encode(e), base64.b64encode(n)
+
+    @api.model
+    def _generate_ec_private_key(self, company_id, name='id_ec', curve='SECP256R1'):
+
+        company = self.env['res.company'].browse(company_id) or self.env.company
```

Comment:
It looks like the company is optional here... I think the company should be either an optional parameter, either a mandatory parameter